zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java

```diff
@@ -3028,6 +3112,77 @@ public void removeAllWatches(String path, WatcherType watcherType, boolean local
         removeWatches(ZooDefs.OpCode.removeWatches, path, null, watcherType, local, cb, ctx);
     }
 
+    /**
+     * <p>
+     * Set a watcher on the given path that: a) does not get removed when triggered (i.e. it stays active
+     * until it is removed); b) optionally applies not only to the registered path but all child paths recursively. This watcher
+     * is triggered for both data and child events. To remove the watcher, use
+     * <tt>removeWatches()</tt> with <tt>WatcherType.Any</tt>
+     * </p>
+     *
+     * <p>
+     * If <tt>recursive</tt> is <tt>false</tt>, the watcher behaves as if you placed an exists() watch and
+     * a getData() watch on the ZNode at the given path.
+     * </p>
+     *
+     * <p>
+     * If <tt>recursive</tt> is <tt>true</tt>, the watcher behaves as if you placed an exists() watch and
+     * a getData() watch on the ZNode at the given path <strong>and</strong> any ZNodes that are children
+     * of the given path including children added later.
+     * </p>
+     *
+     * <p>
+     * NOTE: when there are active recursive watches there is a small performance decrease as all segments
+     * of ZNode paths must be checked for watch triggering.
+     * </p>
+     *
+     * @param basePath the top path that the watcher applies to
+     * @param watcher the watcher
+     * @param recursive if true applies not only to the registered path but all child paths recursively including
+     *                  any child nodes added in the future
+     * @throws InterruptedException If the server transaction is interrupted.
+     * @throws KeeperException If the server signals an error with a non-zero
+     *  error code.
+     */
+    public void addPersistentWatch(String basePath, Watcher watcher, boolean recursive)
```

Comment:
What about using an enum instead of a boolean for 'recursive'?

Reply:
Do you anticipate more types? It would be easy to do.

Reply:
Yes.
We could also call the method 'addWatch' and have enum values like PERSISTENT_RECURSIVE and PERSISTENT_RECURSIVE

Or even better we can use EnumSet and two flags PERSISTENT,RECURSIVE

addWatch(...., EnumSet.of(PERSISTENT,RECURSIVE))

Reply:
The problem with an EnumSet (or a var args) is that it would allow empty sets/arrays which is illegal. I'd rather have a single enum argument.

Reply:
So, I propose adding:

```java
public enum PersistentWatchMode {
        NON_RECURSIVE,
        RECURSIVE
    }
```

And changing the APIs to take this instead of the `boolean`. OK?

Reply:
+1 from me on PersistentWatchMode

Reply:
Hmm - I wonder now if we should rename the method to something more generic like `addWatch(...)` and make the enum `AddWatchMode`. This way we have more flexibility in the future. Thoughts?

Reply:
That's what I suggest above :-)

Reply:
Doh :D 🤣

Reply:
Note: this is actually a pretty big change. I'm going to flow the watch mode through the server and not just make this a client change.

Reply:
we should narrow down the patch to the minimum needed, otherwise it will be hard to merge it soon.
if we have more complex plan for the future and change deeply protocols and server side flows we should discuss on the ML more.

Reply:
So, you're suggesting a client only change? i.e. this enum is merely to avoid using a boolean in the public API?

Reply:
yes
You already did a great work and the client with your changes is  still compatibile at 100% with 3.5 servers in case of not using persistent watches.
If we change the protocol regarding watches (adding some watchtype to the wireprotocol) you will really need a "compatibility flag" on the client configuration or some more complicated handshake with version/feature discovery.

Reply:
Look at the commit I just pushed. This is merely a refinement. I think this is much better actually and we'll thank ourselves in the future. Now, things are more generic "add watch with a mode". The mode is currently only persistent or recursive but we now have room to change in the future. The number/scope of the change is exactly the same as before.